### PR TITLE
DDF-2502: Created a plugin for admins to conditionally add attributes…

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Provides admins with a plug-in for writing conditions which, upon true, apply a set of new attributes to metacards upon ingest. Primarily used for security markings. 

##### Design:
Using a managed service factory, create an instance of `MetacardIngestNetworkPlugin` as a service for each equality rule that must be evaluated. This may cause some overhead and prevents the rules from being strictly ordered by the admin. All logical constructs are limited to those whose additions on attributes are mutually exclusive (i.e. specifying the same attribute to add to the metacard in more than one rule that evaluates to true will result in an undetermined final attribute value). 

#### Who is reviewing it?
@lessarderic 
@shaundmorris 
@brjeter 
@figliold 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[Core APIs](https://github.com/orgs/codice/teams/core-apis)
[Security](https://github.com/orgs/codice/teams/security)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@pklinef 
@stustison

#### How should this be tested? (List steps with links to updated documentation)
* Setup an instance of DDF completely configured with a proper IP/host so other machines can access it (use `dev=true` since cert generation is not necessary). 
* Grab some IPs of local, peer machines
* Configure the plugin with rules based on the collected IPs
* Have the peer machines ingest data and validate the attributes in the rules were assigned accordingly. 

The following bash command will print a machine's out-bound LAN IP address and will be useful during testing: 

```
route get default | grep interface | awk '{print $NF}' | xargs ifconfig | grep "inet " | awk '{print $2}'
```

##### Also validate: 
Scheme, hostname, and context path rules.

#### Any background context you want to provide?
This feature is needed primarily for security markings, which need to be added to the metacard based on its source. 

#### What are the relevant tickets?
[DDF-2502](https://codice.atlassian.net/browse/DDF-2502)
[DDF-2500](https://codice.atlassian.net/browse/DDF-2500)

##### See also: 
[DDF-2577](https://codice.atlassian.net/browse/DDF-2577)

#### Checklist:
- [x] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

### CODE REVIEW

**Comment by Lambeaux** [comment]
Running my formatter on `TestCatalog` in the integration-tests directory had some major effects on code I didn't touch. Has that file not been formatted or do I need a formatter update?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 1]
```diff
@@ -0,0 +1,75 @@
+==== Metacard Ingest Network Plugin
+
+Allows the conditional insertion of new attributes on metacards during ingest based on network information from the ingest request; including IP address and hostname.
+
+For the extent of this section, a 'rule' will refer to a configured, single instance of this plugin.
+
+===== Installing the Metacard Ingest Network Plugin
+
+The Metacard Ingest Network Plugin is installed by default during a standard installation in the catalog application.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: "in the ${ddf-catalog} application."

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 2]
```diff
@@ -0,0 +1,75 @@
+==== Metacard Ingest Network Plugin
+
+Allows the conditional insertion of new attributes on metacards during ingest based on network information from the ingest request; including IP address and hostname.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: "The Metacard Ingest Network Plugin allows.."

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+===== Installing the Metacard Ingest Network Plugin
+
+The Metacard Ingest Network Plugin is installed by default during a standard installation in the catalog application.
+
+===== Configuring the Metacard Ingest Network Plugin
+
+Multiple instances of the plugin can be configured by clicking on its configuration title within the configuration tab of the catalog app. Each instance represents a conditional statement, or a 'rule', that gets evaluated for each ingest request. For any request that meets the configured criteria of a rule, that rule will attempt to transform its list of key-value pairs to become new attributes on all metacards in that request.
+
+The condition piece of the rule is divided into two fields: "Criteria" and "Expected Value". The "Criteria" field features a drop-down list containing the four elements for which equality can be tested:
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: How about "The rule is divided..."

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+contact.contributor-name = John Doe
+contact.contributor-email = john.doe@example.net
+language = English
+language = English, French, German
+security.access-groups = SJ202, SR 101, JS2201
+----
+
+====== Useful Attributes
+
+The following table provides data on some useful attributes that may want to be set by this plugin:
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: maybe "some useful attributes that may commonly be set by this plugin:"

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 5]
```diff
@@ -0,0 +1,75 @@
+==== Metacard Ingest Network Plugin
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: Should probably put a `-` in the file name between `metacard` and `ingest` for consistency. 

Also, I didn't see that this section was getting included in the docs anywhere.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            <Option label="IP Address" value="remoteAddr"/>
+            <Option label="Host Name" value="remoteHost"/>
+            <Option label="Scheme" value="scheme"/>
+            <Option label="Context Path" value="contextPath"/>
+        </AD>
+
+        <AD description="The value that the criteria must equate to for the attribute overrides to occur."
+            name="Expected Value" id="expectedValue" required="true" type="String" default=""/>
+
+        <AD description="New metacard attributes to apply; if an attribute is specified here, it will be placed on the metacard only if the attribute is not already set. The format should be 'key=value' where 'value' can be a comma-separated list of strings. Whitespace is ignored."
```
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/resources/OSGI-INF/metatype/metatype.xml]: Whitespace**s are** ignored?

File: catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        networkPlugin.setExpectedValue(EXPECTED_VALUE);
+        networkPlugin.setNewAttributes(newAttributes);
+
+        assertThat(networkPlugin.getCriteriaKey(), is(CRITERIA_KEY));
+        assertThat(networkPlugin.getExpectedValue(), is(EXPECTED_VALUE));
+        assertThat(networkPlugin.getNewAttributes(), is(newAttributes));
+    }
+
+    @Test
+    public void testPassthroughMethods() throws Exception {
```
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java]: Should we split this into individual tests to make it clear we cover all the methods?

File: catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        assertThat(plugin.processPreDelete(deleteRequest), is(deleteRequest));
+        assertThat(plugin.processPreQuery(queryRequest), is(queryRequest));
+        assertThat(plugin.processPreResource(resourceRequest), is(resourceRequest));
+
+        assertThat(plugin.processPostDelete(deleteResponse), is(deleteResponse));
+        assertThat(plugin.processPostQuery(queryResponse), is(queryResponse));
+        assertThat(plugin.processPostResource(resourceResponse, mock(Metacard.class)),
+                is(resourceResponse));
+
+        verifyZeroInteractions(mockMetacardCondition,
```
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java]: Love all your `verify` calls 👍

File: catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+import ddf.catalog.operation.ResourceRequest;
+import ddf.catalog.operation.ResourceResponse;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.StopProcessingException;
+
+/**
+ * Ensure the rules and config options work correctly for any {@link MetacardIngestNetworkPlugin}.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class MetacardIngestNetworkPluginTest {
```
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java]: If you replace the setters with an update method, we need to make sure to handle and test for the case where the update method is `null`.

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+===== Configuring the Metacard Ingest Network Plugin
+
+Multiple instances of the plugin can be configured by clicking on its configuration title within the configuration tab of the catalog app. Each instance represents a conditional statement, or a 'rule', that gets evaluated for each ingest request. For any request that meets the configured criteria of a rule, that rule will attempt to transform its list of key-value pairs to become new attributes on all metacards in that request.
+
+The condition piece of the rule is divided into two fields: "Criteria" and "Expected Value". The "Criteria" field features a drop-down list containing the four elements for which equality can be tested:
+
+* IP Address
+* Host Name
+* Scheme
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: Should we have a description and/or example for those, especially for _Scheme_  and _Context Path_? Or maybe update the example to include those as well?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+* Context Path
+
+In order for a rule to evaluate to true and the attributes be applied, the value in the "Expected Value" field must be an exact match to the actual value of the selected criteria. For example, if the selected criteria is "IP Address" with an expected value of "192.168.0.1", the rule only evalutes to true for ingest requests coming from "192.168.0.1" and nowhere else.
+
+.Check for IPv6
+IMPORTANT: Verify your system's IP configuration. Rules using "IP Address" may need to be written in IPv6 format.
+
+The key-value pairs within each rule should take the following form: "key = value" where the "key" is the name of the attribute and the "value" is the value assigned to that attribute. Whitespace is ignored unless it is within the key or value. Multi-valued attributes can be expressed in comma-separated format if necessary.
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: Whitespace**s are** ignored?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+.Valid Attribute Assignments
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: **Example of** Valid Attribute Assignments?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+|security.access-individuals
+|Any String
+|yes
+|===
+
+===== Usage Limitations of the Metacard Ingest Network Plugin
+
+* This plugin only works for ingest (create requests).
+* Any attribute that is already set on the metacard will not be overwritten by the plugin.
+* The order of execution is not guaranteed. For any rule configuration where two or more rules add different values for the same attribute, it is unclear what the final value for that attribute will be in the case where more than one of those rules evaluates to true.
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: Should we use _undefined_ instead of unclear?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+Multiple instances of the plugin can be configured by clicking on its configuration title within the configuration tab of the catalog app. Each instance represents a conditional statement, or a 'rule', that gets evaluated for each ingest request. For any request that meets the configured criteria of a rule, that rule will attempt to transform its list of key-value pairs to become new attributes on all metacards in that request.
+
+The condition piece of the rule is divided into two fields: "Criteria" and "Expected Value". The "Criteria" field features a drop-down list containing the four elements for which equality can be tested:
+
+* IP Address
+* Host Name
+* Scheme
+* Context Path
+
+In order for a rule to evaluate to true and the attributes be applied, the value in the "Expected Value" field must be an exact match to the actual value of the selected criteria. For example, if the selected criteria is "IP Address" with an expected value of "192.168.0.1", the rule only evalutes to true for ingest requests coming from "192.168.0.1" and nowhere else.
```
Comment (lessarderic) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: Typo: evalu**a**tes

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 15]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+        final String clientIP = "0:0:0:0:0:0:0:1";
+        final String clientScheme = "https";
+
+        Map<String, Object> networkRule1Properties = getServiceManager().getMetatypeDefaults(
+                symbolicName,
+                factoryPid);
+
+        Map<String, Object> networkRule2Properties = Maps.newHashMap(networkRule1Properties);
+
+        List<Object> attributeAdjustmentsForRule1 = ImmutableList.builder()
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: You should be able to skip the `build` call by doing `ImmutableList.of("security.access-groups = FIFA, NFL");`.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 16]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+                .add("security.access-groups = FIFA, NFL")
+                .build();
+
+        List<Object> attributeAdjustmentsForRule2 = ImmutableList.builder()
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: Same here.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 17]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+                .add("description = None")
+                .add("metacard-tags = ALPHA, BRAVO, resource")
+                // ** Remove below addition when IP gets fixed **
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: ⭐️ `TODO` with ticket number of remove.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 18]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+        networkRule2Properties.put("criteriaKey", "scheme");
+        networkRule2Properties.put("expectedValue", clientScheme);
+        networkRule2Properties.put("newAttributes", attributeAdjustmentsForRule2);
+
+        getServiceManager().createManagedService(factoryPid, networkRule1Properties);
+        getServiceManager().createManagedService(factoryPid, networkRule2Properties);
+
+        getServiceManager().waitForRequiredBundles(symbolicName);
+
+        // ** INGEST **
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: Can we skip the `**` in the comments?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 19]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+        startValidateBody(card1Id).assertThat()
+                .body(hasInitialDescription)
+                .body(hasNewMetacardTags);
+        startValidateBody(card2Id).assertThat()
+                .body(hasNewDescription)
+                .body(hasInitialMetacardTags);
+
+        // ** CLEAN UP **
+        // ** Don't let the conditions spill-over and impact other tests **
+        getServiceManager().stopFeature(true, "catalog-metacardingest-network");
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: ⭐️ This should probably be done in the tear-down method in case the test fails.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 20]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+    }
+
+    private Matcher<Node> hasMetacardAttributeXPath(String attributeName,
+            String... attributeValues) {
+        final String basePathFormat = "/metacard/string[@name=\"%s\"][value=\"%s\"";
+        final String additionalValueFormat = " and value=\"%s\"";
+
+        String xPath = format(basePathFormat, attributeName, attributeValues[0]);
+        for (int i = 1; i < attributeValues.length; i++) {
+            xPath += format(additionalValueFormat, attributeValues[i]);
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: ⭐️ Using both string concatenation _and_ format at the same time seems a little odd and inefficient. We should use one or the other, or simply use a `StringBuilder`.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java [hunk 21]
```diff
@@ -1552,6 +1571,130 @@ public void testIngestPlugin() throws Exception {
...
+                .body(hasNewDescription)
+                .body(hasInitialMetacardTags);
+
+        // ** CLEAN UP **
+        // ** Don't let the conditions spill-over and impact other tests **
+        getServiceManager().stopFeature(true, "catalog-metacardingest-network");
+
+    }
+
+    private ValidatableResponse startValidateBody(String metacardId) {
```
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: This method could be re-used. Could we make the name a little more generic, e.g., `getMetacardValidatableResponse(String source, String metacardId)`, so people know they can re-use it?

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc (hunk 5 above)
Comment (Lambeaux) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: I could do that but then the module name wouldn't match and it would be a convention violation to allow it to in that case; technically I'm already breaking convention adding a dash before `network` in this case but it helps readability. 

If docs don't have to match the original maven module then I'm ok with this.

Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: I don't think it's as important for it to match the module. (I'm fine with being overruled on this though)

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+    /**
+     * Default constructor.
+     */
+    public MetacardIngestNetworkPlugin(KeyValueParser keyValueParser,
+            MetacardServices metacardServices) {
+        this(keyValueParser, metacardServices, new AttributeFactory(), new MetacardCondition());
+    }
+
+    /**
+     * Extended constructor with an {@link AttributeFactory} and {@link MetacardCondition} to provide
```
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: Saves us a little extra blueprint hassle. I think it's fine as is unless there are other compelling reasons to change it.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+
+    private void apply(List<Metacard> metacards, Map<String, Serializable> criteria) {
+        if (metacardCondition.applies(criteria)) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Metacard condition was true; expected {} of {} equals {}",
+                        metacardCondition.getCriteriaKey(),
+                        metacardCondition.getExpectedValue(),
+                        criteria.get(metacardCondition.getCriteriaKey()));
+            }
+            Map<String, String> parsedAttributes = keyValueParser.parsePairsToMap(newAttributes);
```
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: I was trying to avoid having "redundant" class state but yes, this makes sense. It will look a tad strange because we still need the original `List` to support the metatype.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardCondition.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+    private String criteriaKey;
+
+    private String expectedValue;
+
+    public String getCriteriaKey() {
+        return criteriaKey;
+    }
+
+    public void setCriteriaKey(String criteriaKey) {
```
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardCondition.java]: I don't understand the concern. Isn't it no different than if the variables were pulled out into the actual plugin class? I only did this delegation trick with the condition and the plugin to make testing easier.

Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardCondition.java]: Up until now we've never had to worry about plugins being thread-safe.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+        this.metacardServices = metacardServices;
+        this.attributeFactory = attributeFactory;
+        this.metacardCondition = metacardCondition;
+    }
+
+    public String getCriteriaKey() {
+        return metacardCondition.getCriteriaKey();
+    }
+
+    public void setCriteriaKey(String criteriaKey) {
```
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: Need clarification on this one.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public CreateRequest processPreCreate(CreateRequest input) throws StopProcessingException {
+        apply(input.getMetacards(), getClientInfoProperties());
+        return input;
+    }
+
+    //region - Passthrough  methods
+
+    @Override
+    public UpdateRequest processPreUpdate(UpdateRequest input,
```
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: I'm indifferent. Is this okay with you @pklinef ?

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java (hunk 25 above)
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: Ok. Legitimate threading issue. Going to move toward the `component-managed` strategy and make the `MetacardCondition` immutable.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardCondition.java (hunk 24 above)
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardCondition.java]: Issue to be addressed.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java (hunk 7 above)
Comment (Lambeaux) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java]: Maybe if the plugin operated on more than one method. I think it is fine for now. If it is updated in the future then that is probably the right way to go. But currently we only care about ingest.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+                        criteria.get(metacardCondition.getCriteriaKey()));
+            }
+            Map<String, String> parsedAttributes = keyValueParser.parsePairsToMap(newAttributes);
+            metacardServices.setAttributesIfAbsent(metacards, parsedAttributes, attributeFactory);
+        }
+    }
+
+    private Map<String, Serializable> getClientInfoProperties() throws StopProcessingException {
+        Object info = ThreadContext.get(CLIENT_INFO_KEY);
+        if (isFalse(info instanceof Map)) {
```
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: Not really. Just felt overkill in this case since we know `null` will not be an instance of `Map`.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java (hunk 9 above)
Comment (lessarderic) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/test/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPluginTest.java]: Sorry, poorly worded :-(
The `Map` passed in. We've run into that problem in the past where Config Admin would call `update(null)` and our code would throw a NPE.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java (hunk 19 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: Then we will need to start the feature at the beginning of the test method; else it won't be running when we need it.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java (hunk 17 above)
Comment (Lambeaux) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: This needs fixing before this gets merged.

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc (hunk 11 above)
Comment (Lambeaux) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: While it may not be intuitive, `whitespace` can be used as a plural form. It can refer to a single symbol or a collection (such as outer space is the collective sum of the space outside earth, not a single portion).

File: distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc (hunk 5 above)
Comment (Lambeaux) [inline distribution/docs/src/main/resources/_contents/_plugins/metacardingest-network-contents.adoc]: I think it's a reasonable expectation that programmers will think the file name is the same as the module; that's at least according to the principle of least astonishment. Would be a shame if it didn't appear on a search for `metacardingest` literally.

File: catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java (hunk 26 above)
Comment (pklinef) [inline catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/MetacardIngestNetworkPlugin.java]: Default methods can cause issues with OSGi service and backwards compatibility.  If the interface is shared between bundles then both bundles need to be compiled with the same version of the interface.  That means that integrators would need to recompile their bundles any time the default methods are changed.  I would recommend an abstract class instead which would do not cause a backward compatibility issue in OSGi.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java (hunk 19 above)
Comment (lessarderic) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestCatalog.java]: Looks like trying to uninstall a feature that's not been install will throw an exception, but it may not matter. We could always surround the `stopFeature()` in the tear-down method with a try/catch and ignore all exceptions. That way it wouldn't matter whether the feature was started or not in a test.

What I don't like about the finally block option is that we can easily forget to add the same logic to any future tests we write that need the feature to be running.

**Comment by Lambeaux** [comment]
Still need to address test failures as a result of switching to `component-managed` for the plugin.

**Comment by Lambeaux** [comment]
Ready for a hero after a final approval from @pklinef and @stustison - all comments addressed.

**Comment by brjeter** [comment]
👍  Heroed successfully
